### PR TITLE
IO: Refactor directory layout

### LINKED ISSUE #8691: Refactor the IO related directory
## Enhancement

split `dbms/src/IO` and `dbms/src/Encryption` into

- BaseFile: classes that manipulate files directly
- Encryption: encryption and checksum
- Compression: compression
- IO: all kinds of buffers, the objects that we manipulate directly when we want to do IO.

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: close #8691

Problem Summary:

### What is changed and how it works?

The last PR for #8691

1. add `FileProvider/` directory, which contains FileProvider and all kinds of builders.
2. remove `Util/` directory, move all files under `IO/`
3. remove `CompressedReadBufferFromFileProvider` which is nearly the same as `CompressedReadBufferFromFile`, and add `CompressedReadBufferFromFileBuilder`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by CalvinNeo — APPROVED** [review]
lgtm

File: dbms/src/IO/Compression/CompressedReadBufferFromFile.cpp [hunk 1]
```diff
@@ -12,28 +12,36 @@
...
+template <bool has_legacy_checksum>
+bool CompressedReadBufferFromFileImpl<has_legacy_checksum>::nextImpl()
 {
     size_t size_decompressed;
     size_t size_compressed_without_checksum;
     size_compressed = this->readCompressedData(size_decompressed, size_compressed_without_checksum);
     if (!size_compressed)
         return false;
 
-    assert(size_decompressed > 0 && size_compressed_without_checksum > 0);
```
Comment (JaySon-Huang) [inline dbms/src/IO/Compression/CompressedReadBufferFromFile.cpp:36]: why `size_compressed_without_checksum > 0` is removed?

Comment (Lloyd-Pottiger) [reply dbms/src/IO/Compression/CompressedReadBufferFromFile.cpp:36]: Not necessary, this file is rename from `CompressedReadBufferFromFileProvider.cpp`

The assert is used to pass tidy, `size_compressed_without_checksum > 0` is not necessary.

File: dbms/src/IO/FileProvider/CompressedReadBufferFromFileBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    const ReadLimiterPtr & read_limiter_,
+    size_t buf_size)
+{
+    auto file_in = ReadBufferFromRandomAccessFileBuilder::buildPtr(
+        file_provider,
+        path,
+        encryption_path,
+        buf_size,
+        read_limiter_);
+    return std::make_unique<CompressedReadBufferFromFileImpl<true>>(std::move(file_in));
```
Comment (JaySon-Huang) [inline dbms/src/IO/FileProvider/CompressedReadBufferFromFileBuilder.cpp:36]:
```suggestion
    // with legacy checksum in CompressedReadBuffer
    return std::make_unique<CompressedReadBufferFromFileImpl<true>>(std::move(file_in));
```

**Review by JaySon-Huang — COMMENTED** [review]
Rest LGTM

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**